File: BE/src/main/java/com/sugyo/domain/game/domain/GameSessionContext.java
package com.sugyo.domain.game.domain;

import lombok.Getter;
import lombok.ToString;

import java.time.Instant;
import java.util.concurrent.atomic.AtomicInteger;

import static com.sugyo.domain.game.domain.GameState.PLAYING;

@Getter
@ToString
public class GameSessionContext {

    private final String userId;
    private final Long musicId;
    private final String webSocketSessionId;

    private final AtomicInteger score;      // 누적 점수
    private final AtomicInteger combo;      // 현재 콤보
    private final AtomicInteger perfectCount; // Perfect 판정 횟수
    private final AtomicInteger goodCount;    // Good 판정 횟수
    private final AtomicInteger missCount;    // Miss 판정 횟수

    private volatile GameState gameState;   // 게임 상태 (volatile로 가시성 보장)
    private volatile Instant lastActivityTime; // 마지막 활동 시간 (타임아웃 처리용)


    public GameSessionContext(String userId, Long songId, String webSocketSessionId){
        this.userId = userId;
        this.musicId = songId;
        this.webSocketSessionId = webSocketSessionId;

        this.score = new AtomicInteger(0);
        this.combo = new AtomicInteger(0);
        this.perfectCount = new AtomicInteger(0);
        this.goodCount = new AtomicInteger(0);
        this.missCount = new AtomicInteger(0);

        this.gameState = PLAYING;
        this.lastActivityTime = Instant.now();
    }

    public boolean isPlaying(){
        return this.gameState == PLAYING;
    }

    public void applyJudgment(int points, Judgment judgment){
        this.score.addAndGet(points);
        recordActivity();

        switch (judgment){
            case PERFECT -> {
                this.combo.incrementAndGet();
                this.perfectCount.incrementAndGet();
            }
            case GOOD -> {
                this.combo.incrementAndGet();
                this.goodCount.incrementAndGet();
            }
            case MISS -> {
                this.combo.set(0);
                this.missCount.incrementAndGet();
            }
        }
    }

    public void changeState(GameState newState){
        this.gameState = newState;
        recordActivity();
    }

    private void recordActivity(){
        this.lastActivityTime = Instant.now();
    }
}

File: app/src/main/java/com/ssafy/a602/game/play/judge/GameSessionContext.kt
package com.ssafy.a602.game.play.judge

import java.time.Instant
import java.util.concurrent.atomic.AtomicInteger

data class GameSessionContext(
    val userId: String,
    val musicId: Long,
    val webSocketSessionId: String,
    val score: AtomicInteger = AtomicInteger(0),
    val combo: AtomicInteger = AtomicInteger(0),
    val perfectCount: AtomicInteger = AtomicInteger(0),
    val goodCount: AtomicInteger = AtomicInteger(0),
    val missCount: AtomicInteger = AtomicInteger(0),
    var gameState: GameState = GameState.PLAYING,
    var lastActivityTime: Instant = Instant.now(),
    val lastNoteTimestamp: Float
) {
    
    constructor(userId: String, songId: Long, sessionId: String, lastNoteTimestamp: Float) : this(
        userId = userId,
        musicId = songId,
        webSocketSessionId = sessionId,
        score = AtomicInteger(0),
        combo = AtomicInteger(0),
        perfectCount = AtomicInteger(0),
        goodCount = AtomicInteger(0),
        missCount = AtomicInteger(0),
        gameState = GameState.PLAYING,
        lastActivityTime = Instant.now(),
        lastNoteTimestamp = lastNoteTimestamp
    )

    fun isPlaying(): Boolean {
        return this.gameState == GameState.PLAYING
    }

    fun applyJudgment(points: Int, judgment: Judgment) {
        this.score.addAndGet(points)
        recordActivity()

        when (judgment) {
            Judgment.PERFECT -> {
                this.combo.incrementAndGet()
                this.perfectCount.incrementAndGet()
            }
            Judgment.GOOD -> {
                this.combo.incrementAndGet()
                this.goodCount.incrementAndGet()
            }
            Judgment.MISS -> {
                this.combo.set(0)
                this.missCount.incrementAndGet()
            }
        }
    }

    fun changeState(newState: GameState) {
        this.gameState = newState
        recordActivity()
    }

    private fun recordActivity() {
        this.lastActivityTime = Instant.now()
    }
}

enum class GameState {
    PLAYING, PAUSED, FINISHED
}
